### PR TITLE
Further work on applying cmake GenerateExportHeader

### LINKED ISSUE #27061: Use CMake GenerateExportHeader instead of doing this manually.
- [x] Algorithms #28335
- [x] API #28378
- [x] Beamline #27113
- [x] Catalog #27113
- [x] Crystal #28378
- [x] CurveFitting #28378
- [ ] DataHandling
- [ ] DataObjects
- [ ] EventParallelLoader
- [ ] Geometry 
- [x] HistogramData #27113
- [x] ICat #27113
- [x] Indexing #27102
- [ ] Kernel
- [x] LiveData #27113
- [ ] MDAlgorithms & MPIAlgorithms
- [x] Muon #27113
- [ ] Nexus & NexusGeometry
- [ ] Parallel
- [ ] PythonInterface 
- [ ] RemoteAlgorithms & RemoteJobManagers
- [ ] ScriptRepository & SINQ
- [ ] TestHelpers
- [x] Types #27102
- [ ] WorkflowAlgorithms & [`class_maker`]

### PR DESCRIPTION
**Pull request by DannyHindson**
**Description of work.**

Apply change to insert export macros into class definition using cmake GenerateExportHeader
feature to 3 more projects: API, Crystal and CurveFitting (see linked issue for previous work on this)
Have replaced all occurrences of DllExport with the cmake macro except in these places:

a) class templates - the definitions of class template member functions need to be in the
same unit as the declaration so using dllimport gives a compilation error in the client code
(C2491 in VS) eg SingleValueParameter, SingleValueParameterParser
b) headers which are in a project but aren't included by any .cpp units in the project
eg IBackgroundFunction.h, ICatalogInfoService.h. In these cases I've removed the DLLExport
attribute completely

For the API target there were a couple of special cases\extras:

API has some conditional use of the extern keyword to control instantiation of template classes. This has been inserted into the automatically created dllconfig.h using a cmake switch that allows some custom content to be appended to the end of the file. Few of the targets require this so I've added a boolean parameter to the wrapper function to handle this GENERATE_MANTID_EXPORT_HEADER

I also had to add explicit include directories to PythonInterfaceCore because it imports an API target
header (in the unit ExtractWorkspace.h) without having API as a dependency.

**To test:**

Assuming the change passes the various Jenkins checks this just needs a code review

Refs #27061

*This does not require release notes* because it's an internal technical change with no external change to Mantid functionality

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Crystal/inc/MantidCrystal/SCDCalibratePanels.h [hunk 1]
```diff
@@ -7,6 +7,7 @@
 #pragma once
 
 #include "MantidAPI/Algorithm.h"
+#include "MantidCrystal/DllConfig.h"
```
Comment (ConorMFinn) [inline Framework/Crystal/inc/MantidCrystal/SCDCalibratePanels.h]: I've noticed a couple of these there `DllConfig.h` is added but nothing else is changed, is this intentional?

**Review by cailafinn — CHANGES_REQUESTED** [review]
Builds and launches properly, just a couple of things I found while reviewing.

File: Framework/Crystal/inc/MantidCrystal/SCDCalibratePanels.h (hunk 1 above)
Comment (DannyHindson) [reply Framework/Crystal/inc/MantidCrystal/SCDCalibratePanels.h]: No that shouldn't be there. I'll remove

File: Framework/API/test/AlgorithmPropertyTest.h [hunk 2]
```diff
@@ -6,7 +6,7 @@
 // SPDX - License - Identifier: GPL - 3.0 +
 #pragma once
 
-#include "MantidKernel/System.h"
+//#include "MantidKernel/System.h"
```
[the remark this answers is not in the document]
Comment (DannyHindson) [reply Framework/API/test/AlgorithmPropertyTest.h]: Thanks for spotting. I'll remove

**Comment by DannyHindson** [comment]
@ConorMFinn Thanks a lot for ploughing through all of this. Have just made the required changes

**Review by cailafinn — APPROVED** [review]
Looks good to me now. Nice work :+1:

**Pull request merged**